src/lib/parser.c

```diff
@@ -840,35 +777,239 @@ static RdbStatus listListpackItem(RdbParser *p, BulkInfo *lpInfo) {
 
 /* return either RDB_STATUS_OK or RDB_STATUS_ERROR */
 static RdbStatus listZiplistItem(RdbParser *p, BulkInfo *ziplistBulk) {
-    int ret;
 
-    if (p->elmCtx.key.handleByLevel == RDB_LEVEL_STRUCT) {
-        ret = ziplistValidateIntegrity(ziplistBulk->ref, ziplistBulk->len, p->deepIntegCheck, NULL, NULL);
+    int ret = ziplistValidateIntegrity(ziplistBulk->ref, ziplistBulk->len, p->deepIntegCheck, NULL, NULL);
 
-        if (unlikely(!ret)) {
-            RDB_reportError(p, RDB_ERR_ZIP_LIST_INTEG_CHECK, "elementZiplist(): Ziplist integrity check failed");
-            return RDB_STATUS_ERROR;
-        }
+    if (unlikely(!ret)) {
+        RDB_reportError(p, RDB_ERR_LIST_ZL_INTEG_CHECK, "listZiplistItem(): Ziplist integrity check failed");
+        return RDB_STATUS_ERROR;
+    }
 
+    if (p->elmCtx.key.handleByLevel == RDB_LEVEL_STRUCT) {
         registerAppBulkForNextCb(p, ziplistBulk);
         CALL_HANDLERS_CB(p, NOP, RDB_LEVEL_STRUCT, rdbStruct.handleListZL, ziplistBulk->ref);
+        return RDB_STATUS_OK;
+    }
+
+    unsigned char *offsetZL = ziplistIndex(ziplistBulk->ref, 0);
+    while (offsetZL != NULL) {
+        unsigned char *item, endChar, *pEndChar;
+        unsigned int itemLen;
+        long long itemVal;
+        BulkInfo *binfo;
+
+        ziplistGet(offsetZL, &item, &itemLen, &itemVal);
+        offsetZL = ziplistNext(ziplistBulk->ref, offsetZL);
+
+        if (!(binfo = allocBulkForEmbeddedString(p, item, itemLen, itemVal, &endChar, &pEndChar)))
+            return RDB_STATUS_ERROR;
+
+        registerAppBulkForNextCb(p, binfo);
+        CALL_HANDLERS_CB(p,
+                         *pEndChar = endChar, // finalize by restore modified char
```

Comment:
maybe wrapping this assignment in a function would make it more readable?
i.e. calling a function and passing these two vars as arguments.
it'll mean it has a name, and a better place to write a bigger comment without having to repeat it.
i don't recall if i already suggested it, and maybe it's impossible...